### ISSUE TITLE
How can i apply transforms ex. (toggle visibility) to parent `StateObject`, but not to child? How to separate them otherwise?

### ISSUE THREAD

**Issue by rtviii** [molstar/molstar#1186]
I apologize as usual if my description of this is misconceived in terms of state representations:
I load a structure, create a separate(? so i thought) ligand object inside of it and now want to toggle the structure's visibility.

Download struct:

```ts
    const data = await this.ctx.builders.data.download({ url: `https://files.rcsb.org/download/${rcsb_id.toUpperCase()}.cif` }, { state: { isGhost: true } });
    const trajectory = await this.ctx.builders.structure.parseTrajectory(data, "mmcif");
    // i'm still not sure where to stick this ref to save it on creation... Tips?
    const structRef = `structure-${rcsb_id}`; 
    const st = await this.ctx.builders.structure.hierarchy.applyPreset(trajectory, "default");
    this.ctx.managers.structure.component.setOptions({ ...this.ctx.managers.structure.component.state.options, ignoreLight: true });
```

Create ligand:
```ts
      ...
      const struct = structures[0];
      const update = this.ctx.build();

      const ligand = MS.struct.filter.first([
        MS.struct.generator.atomGroups({
          'residue-test': MS.core.rel.eq([MS.struct.atomProperty.macromolecular.label_comp_id(), chemicalId]),
          'group-by': MS.core.str.concat([MS.struct.atomProperty.core.operatorName(), MS.struct.atomProperty.macromolecular.residueKey()])
        })
      ]);

// I also tried .to(this.ctx.state.data.root) thinking that would make the new ligand object "separate" from the parent struct, but get "no suitable parent for "ligandSel" found" errors for further selection
      const group1 = update.to(struct.structureRef.cell) 
.group(StateTransforms.Misc.CreateGroup, { label: `${chemicalId} Ligand Group` }, { ref: 'ligand' });
      const ligandSel = group1.apply(StateTransforms.Model.StructureSelectionFromExpression, { label: `${chemicalId} Ligand`, expression: ligand }, { ref: 'ligandSel' });
      ligandSel.apply(StateTransforms.Representation.StructureRepresentation3D, createStructureRepresentationParams(this.ctx, struct.structureRef.cell.obj?.data, { type: 'ball-and-stick' }), { ref: 'ligandBallAndStick' });

      await PluginCommands.State.Update(this.ctx, { state: this.ctx.state.data, tree: update });
      
      const compiled = compile<StructureSelection>(ligand);
      const selection = compiled(new QueryContext(struct.structureRef.cell.obj?.data!));
      let loci = StructureSelection.toLociWithSourceUnits(selection);
```

Then, as suggested in #1133 i iterate over my components trying to find the structure only (i still don't know how to save the ref on creation -- maybe that's the remedy?):
```ts
    var sought_ref = ""
    for (const s of this.ctx.managers.structure.hierarchy.currentComponentGroups) {
      for (var component of s) {
        console.log(component.structure.cell.obj?.data.label)
        console.log(component.cell.obj?.label)
        if (component.cell.obj?.label == "Polymer") {
          sought_ref = component.cell.sourceRef
        }
        console.log(component.cell.sourceRef)
      }
    }

    await PluginCommands.State.ToggleVisibility(this.ctx, { state: this.ctx.state.data, ref: sought_ref });
```

But this apparently toggles visibility for all components... Seeming as different objects in the components tray, i thought maybe i should create my ligand etc. with `update.to(root)`, but that still toggles visibility on everything with the above ref rummaging. Any tips?

### tldr:
- I create ligand with a structure as a context and want to toggle the structure's visibility and apply other transforms, but can't find a way to keep the ligand unaffected by those "parent" transforms. 

- Don't know how to save structure ref on creation so might be getting wrong refs. 

Thank you kindly.

![image](https://github.com/molstar/molstar/assets/34076599/ffafed24-445c-4853-922e-9f93aca4d1a1)

https://github.com/molstar/molstar/issues/1133

**Comment by dsehnal** [comment]
- The `const st = await this.ctx.builders.structure.hierarchy.applyPreset(trajectory, "default");` has selectors for the model, structure and the individual representations, 
- You should likely create your ligand selection as child of `st.structure`
- You can use `setSubtreeVisibility` call directly (`true` means "hide" because this sets `isHidden` property of `Transform.State` -- could have done a better job with naming here :))

**Comment by rtviii** [comment]
For people later:

What worked for me was using `setSubtreeVisibility` the following way:
```ts
  async toggle_visibility_by_ref(representation: any, on_off: boolean) {
    // "representation" is saved on structure creation: const st = await this.ctx.builders.structure.hierarchy.applyPreset(trajectory, "default");
    if (representation['components'] === undefined) { return }
    setSubtreeVisibility(this.ctx.state.data, representation['components']['polymer']['ref'], on_off)
  }
```

This works even if you add more than one representation to the structure. You have to save the `ref` or rather (the `representation`) object on creation. This was a particularly useful bit of exchange for me because i now know how (where in the object) `ctx.builders.structure.hierarchy.applyPreset` yields the refs so i can save them for manipulating created structure precisely.
```ts
  async download_struct(rcsb_id: string, clear?: boolean): Promise<{ ctx: MolstarRibxz, struct_representation: any }> {

    const data = await this.ctx.builders.data.download({ url: `https://files.rcsb.org/download/${rcsb_id.toUpperCase()}.cif` }, { state: { isGhost: true } });
    const trajectory = await this.ctx.builders.structure.parseTrajectory(data, "mmcif");
    
    const st = await this.ctx.builders.structure.hierarchy.applyPreset(trajectory, "default");
    st?.structure.ref // <<<<<
    st?.model.ref     // <<<<<


    return { ctx: this, struct_representation: st?.representation }
  }
  ```
  
  Thanks as always, David.
